framework/src/tinyalsa/tinyalsa.c

```diff
@@ -585,33 +592,92 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 #endif
 
 	/* If device is not yet started, start now! */
-	if ((!pcm->running) && (pcm_start(pcm) < 0)) {
+	if ((!pcm->draining) && (!pcm->running) && (pcm_start(pcm) < 0)) {
 		return -errno;
-	} else {
-		pcm->running = 1;
 	}
 
-	/* Wait for deque message from kernel */
-	size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
-	if (size != sizeof(msg)) {
-		/* Interrupted by a signal? What to do? */
-		return oops(pcm, EINTR, "Interrupted while waiting for deque message from kernel");
-	}
-	if (msg.msgId == AUDIO_MSG_DEQUEUE) {
-		apb = (struct ap_buffer_s *)msg.u.pPtr;
-		/* Copy data to user buffer */
-		memcpy(data, apb->samp, apb->nbytes);
-		/* Enque buffer for next read opertion */
-		nbytes = apb->nbytes;
-		apb->nbytes = 0;
-		apb->curbyte = 0;
-		apb->flags = 0;
-		bufdesc.u.pBuffer = apb;
-		if (ioctl(pcm->fd, AUDIOIOC_ENQUEUEBUFFER, (unsigned long)&bufdesc) < 0) {
-			return oops(pcm, errno, "failed to enque buffer after read");
+	/* Read data */
+	if (pcm->next_buf != NULL) {
+		/* We have a pending buffer with some data in it. */
+		apb = pcm->next_buf;
+
+		if (nbytes < pcm->next_size) {
+			/* User wants less data than we have in buffer.
+			   Copy the requested amount of data to user buffer and return */
+			memcpy(data, apb->samp + pcm->next_offset, nbytes);
+			pcm->next_size -= nbytes;
+			pcm->next_offset += nbytes;
+		} else {
+			/* User wants more data than we have in buffer.
+			   Copy available data to user buffer, enqueue buffer and return */
+			memcpy(data, apb->samp + pcm->next_offset, pcm->next_size);
+
+			nbytes = pcm->next_size;
+			pcm->next_size = 0;
+			pcm->next_offset = 0;
+			pcm->next_buf = NULL;
+
+			apb->nbytes = 0;
+			apb->curbyte = 0;
+			apb->flags = 0;
+			bufdesc.u.pBuffer = apb;
+			if (ioctl(pcm->fd, AUDIOIOC_ENQUEUEBUFFER, (unsigned long)&bufdesc) < 0) {
+				return oops(pcm, errno, "failed to enque buffer after read");
+			}
 		}
+
 	} else {
-		return oops(pcm, EINTR, "Recieved unexpected msg (id = %d) while waiting for deque message from kernel", msg.msgId);
+		/* We got here because we dont have any pending buffers */
+		/* Wait for deque message from kernel */
+		if (pcm->draining) {
+			/* When we are in draining state, we will check for buffers already in the msg queue
+			   If there is no buffer on the queue, we return immediately */
+			clock_gettime(CLOCK_REALTIME, &st_time);
+			size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
+			if (size != sizeof(msg)) {
+				pcm->draining = 0;
+				pcm->next_size = 0;
+				pcm->next_offset = 0;
+				pcm->next_buf = NULL;
+				return oops(pcm, ENODATA, "no data available");
+			}
+		} else {
+			size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
+			if (size != sizeof(msg)) {
+				/* Interrupted by a signal? What to do? */
+				return oops(pcm, EINTR, "Interrupted while waiting for deque message from kernel");
+			}
+		}
+		if (msg.msgId == AUDIO_MSG_DEQUEUE) {
```

Comment:
msg seems that can not be reached here.

Reply:
Sorry, I don't understand why. 
I think, msg will come from either line 636 or line 645.

Reply:
My miss, NVM